processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -346,18 +349,33 @@ public void run()
           // PerPrototypeClassState.specializationStates. But it might be that nobody ever hits even the current
           // maxSpecializations limit, so implementing cache eviction is an unnecessary complexity.
           specialized = perPrototypeClassState.prototypeClass.newInstance();
-          LOG.warn(
-              "SpecializationService couldn't make more than [%d] specializations. "
-              + "Not doing specialization for runtime shape[%s] and class remapping[%s], using the prototype class[%s]",
-              maxSpecializations,
-              specializationId.runtimeShape,
-              specializationId.classRemapping,
-              perPrototypeClassState.prototypeClass
-          );
+          if (!maxSpecializationsWarningEmitted.get() && maxSpecializationsWarningEmitted.compareAndSet(false, true)) {
+            LOG.warn(
+                "SpecializationService couldn't make more than [%d] specializations. " +
+                "Not doing specialization for runtime shape[%s] and class remapping[%s], using the prototype class[%s]",
+                maxSpecializations,
+                specializationId.runtimeShape,
+                specializationId.classRemapping,
+                perPrototypeClassState.prototypeClass
+            );
+          }
         } else if (fakeSpecialize) {
           specialized = perPrototypeClassState.prototypeClass.newInstance();
+          LOG.info(
```

Comment:
My perception is that nobody runs anything with DEBUG logging level, because many third-party libraries generate way too much debug output. So debug logging in Druid = no logging. Currently there is a hard limit on how many such messages you could see during the JVM lifetime: 1000, it's not very much, so I would keep it INFO.